strawberryfields/utils.py

```diff
@@ -673,3 +691,334 @@ def f_proxy(*args):
+    vectorized = np.reshape(transposed, [shape[0]**(dims//4)]*4)
+    transposed_back = np.einsum('abcd -> acbd', vectorized)
+
+    return transposed_back
+
+
+def _unvectorize_dm(tensor, num_subsystems):
+    """Given a tensor with 4 indices, each of dimension :math:`D^N`, return the unvectorized
+    tensor with 4N indices of dimension D each.
+
+    This is the inverse of the procedure given by :func:`_vectorize_dm`.
+
+    Args:
+        tensor (array): a tensor with :math:`4` indices of dimension :math:`D^N`
+
+    Returns:
+        array: a tensor with :math:`4N` indices of dimension :math:`D` each
+
+    Raises:
+        ValueError: if the input tensor's dimensions are not all equal or if the number
+            of its indices is not 4.
+    """
+    dims = tensor.ndim
+
+    if dims != 4:
+        raise ValueError('tensor must have 4 indices, but it has {dims} indices'.format(dims=dims))
+
+    shape = tensor.shape
+
+    if len(set(shape)) != 1:
+        raise ValueError('tensor indices must have all the same dimension, but tensor has shape {shape}'.format(shape=shape))
+
+    transposed = np.einsum('abcd -> acbd', tensor)
+    unvectorized = np.reshape(transposed, [int(shape[0]**(1/num_subsystems))]*(4*num_subsystems))
+    transposed_back = np.einsum(unvectorized, [int(n) for n in np.arange(4*num_subsystems).reshape((2*num_subsystems, 2)).T.reshape([-1])])
+
+    return transposed_back
+
+
+def _interleaved_identities(num_subsystems: int, cutoff_dim: int):
+    """Returns the tensor product of ``num_subsystems`` copies of the identity matrix with the indices interleaved.
+
+    For example, when ``num_subsystems=3``, this function creates the tensor product of
+    three identity matrices :math:`I_{ijklmn}` (where :math:`ij`, :math:`kl`, and :math:`mn`
+    are the three sets of indices), and returns the tensor :math:`I_{ikmjln}`.
+
+    Args:
+        num_subsystems (int): number of subsystems to consider
+        cutoff_dim (int): the Fock basis truncation
+
+    Returns:
+        array: the resulting interleaved identity tensor product
+    """
+    I = np.identity(cutoff_dim)
+    for _ in range(1, num_subsystems):
+        I = np.tensordot(I, np.identity(cutoff_dim), axes=0)
+
+    return np.einsum(I, [int(n) for n in np.arange(2*num_subsystems).reshape((2, num_subsystems)).T.reshape([-1])])
+
+
+def _engine_with_CJ_cmd_queue(engine, cutoff_dim: int):
+    """Doubles the number of modes of an engine objects and prepends to its command queue
+    the operation that creates the interleaved identities ket.
+
+    Here, CJ is from Choi-Jamiolkowski, as under the hood we are expliting the CJ isomorphism.
+
+    Args:
+        engine (Engine): a Strawberry Fields engine instance
+        cutoff_dim (int): the Fock basis truncation
+
+    Returns:
+        strawberryfields.engine.Engine: returns the same Engine instance, with the number of subsystems
+        increased and the interleaved identities ket prepended to the queue
```

Comment:
Just double checking, but because there is no deepcopy here, this is an 'in-place' engine operation?

Reply:
Hi Josh, I'm not sure I understand your question. Could you rephrase?

Reply:
in this function, even though the engine is returned, it is also modified directly - so even if nothing was returned, the passed engine will still be modified.

I see that further down on line 875 where it is being used, you are using deepcopy, so it should be an issue there :slightly_smiling_face:

Reply:
Yeah, I didn't want to mess up the current Engine, so I work on a deepcopy.